src/libfsm/intersect.c

```diff
@@ -65,3 +72,317 @@ fsm_intersect(struct fsm *a, struct fsm *b)
 	return NULL;
 }
 
+struct pair {
+	struct fsm_state *a;
+	struct fsm_state *b;
+	struct fsm_state *comb;
+};
+
+static int cmp_pair(const void *a, const void *b)
+{
+	const struct pair *pa = a, *pb = b;
+	ptrdiff_t delta;
+	delta = pa->a - pb->a;
+	if (delta < 0) {
+		return -1;
+	}
+	
+	if (delta > 0) {
+		return 1;
+	}
+
+	delta = pa->b - pb->b;
+	if (delta < 0) {
+		return -1;
+	}
+
+	if (delta > 0) {
+		return 1;
+	}
+
+	return 0;
+}
+
+struct pair_pool {
+	struct pair_pool *next;
+	struct pair items[PAIR_POOL_SIZE];
+};
+
+struct bywalk_arena {
+	struct pair_pool *head;
+	size_t top;
+
+	struct fsm *new;
+	struct set *states;
+};
+
+static void
+free_bywalk(struct bywalk_arena *A)
+{
+	struct pair_pool *p, *next;
+
+	if (A->states) {
+		set_free(A->states);
+	}
+
+	if (A->new) {
+		fsm_free(A->new);
+	}
+
+	for (p = A->head; p != NULL; p = next) {
+		next = p->next;
+		free(p);
+	}
+}
+
+static struct pair *alloc_pair(struct bywalk_arena *A)
+{
+	static const struct pair zero;
+
+	struct pair *item;
+	struct pair_pool *pool;
+
+	if (A->head == NULL) {
+		goto new_pool;
+	}
+
+	if (A->top >= (sizeof A->head->items / sizeof A->head->items[0])) {
+		goto new_pool;
+	}
+
+	item = &A->head->items[A->top++];
+	*item = zero;
+	return item;
+
+new_pool:
+	pool = malloc(sizeof *pool);
+	if (pool == NULL) {
+		return NULL;
+	}
+
+	pool->next = A->head;
+	A->head = pool;
+	A->top = 1;
+
+	item = &pool->items[0];
+	*item = zero;
+	return item;
+}
+
+static struct pair *new_pair(struct bywalk_arena *A, struct fsm_state *a, struct fsm_state *b)
+{
+	struct pair lkup, *p;
+	struct fsm_state *combined;
+
+	lkup.a = a;
+	lkup.b = b;
+
+	assert(A->states);
+
+	p = set_contains(A->states, &lkup);
+	if (p != NULL) {
+		return p;
+	}
+
+	p = alloc_pair(A);
+	if (p == NULL) {
+		return NULL;
+	}
+
+	combined = fsm_addstate(A->new);
+	if (combined == NULL) {
+		return NULL;
+	}
+	combined->equiv = NULL;
+
+	p->a = a;
+	p->b = b;
+	p->comb = combined;
```

Comment:
`comb` or `combined`? If it's worth spelling out, it's worth it for all instances.

Reply:
(Personally I'd call it `q` because of my electronics background, but I thought `comb` was fine)